pkg/engine/planner/physical/optimizer.go

```diff
@@ -0,0 +1,183 @@
+package physical
+
+import (
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+)
+
+// A rule is a tranformation that can be applied on a Node.
+type rule interface {
+	// apply tries to apply the transformation on the node.
+	// It returns a boolean indicating whether the transformation has been applied.
+	apply(Node) bool
+}
+
+// removeNoopFilter is a rule that removes Filter nodes without predicates.
+type removeNoopFilter struct {
+	plan *Plan
+}
+
+// apply implements rule.
+func (r *removeNoopFilter) apply(node Node) bool {
+	changed := false
+	switch node := node.(type) {
+	case *Filter:
+		if len(node.Predicates) == 0 {
+			r.plan.removeNode(node)
+			changed = true
+		}
+	}
+	return changed
+}
+
+var _ rule = (*removeNoopFilter)(nil)
+
+// predicatePushdown is a rule that moves down filter predicates to the scan nodes.
+type predicatePushdown struct {
+	plan *Plan
+}
+
+// apply implements rule.
+func (r *predicatePushdown) apply(node Node) bool {
+	changed := false
+	switch node := node.(type) {
+	case *Filter:
+		for i := 0; i < len(node.Predicates); i++ {
+			if ok := r.applyPredicatePushdown(node, node.Predicates[i]); ok {
+				changed = true
+				// remove predicates that have been pushed down
+				node.Predicates = append(node.Predicates[:i], node.Predicates[i+1:]...)
+				i--
+			}
+		}
+	}
+	return changed
+}
+
+func (r *predicatePushdown) applyPredicatePushdown(node Node, predicate Expression) bool {
+	switch node := node.(type) {
+	case *DataObjScan:
+		if canApplyPredicate(predicate) {
+			node.Predicates = append(node.Predicates, predicate)
+			return true
+		}
+		return false
+	}
+	for _, child := range r.plan.Children(node) {
+		if ok := r.applyPredicatePushdown(child, predicate); !ok {
+			return ok
+		}
+	}
+	return true
+}
+
+func canApplyPredicate(predicate Expression) bool {
+	switch pred := predicate.(type) {
+	case *BinaryExpr:
+		return canApplyPredicate(pred.Left) && canApplyPredicate(pred.Right)
+	case *ColumnExpr:
+		return pred.ColumnType == types.ColumnTypeBuiltin || pred.ColumnType == types.ColumnTypeMetadata
+	case *LiteralExpr:
+		return true
+	default:
+		return false
+	}
+}
+
+var _ rule = (*predicatePushdown)(nil)
+
+// limitPushdown is a rule that moves down the limit to the scan nodes.
+type limitPushdown struct {
+	plan *Plan
+}
+
+// apply implements rule.
+func (r *limitPushdown) apply(node Node) bool {
+	switch node := node.(type) {
+	case *Limit:
+		return r.applyLimitPushdown(node, node.Limit)
+	}
+	return false
+}
+
+func (r *limitPushdown) applyLimitPushdown(node Node, limit uint32) bool {
+	switch node := node.(type) {
+	case *DataObjScan:
+		node.Limit = max(node.Limit, limit)
```

Comment:
I think it's worth adding a comment here to explain to readers why this is important (i.e., properly handling the case when a DataObjScan is reachable from two different Limit nodes)

Reply:
Agree, this wasn't initially clear to me.